engineering-plans/0002-subgraph-schema-merging.md

```diff
@@ -0,0 +1,118 @@
+# PLAN-0001: Subgraph Schema Merging for Subgraph Composition
+
+<dl>
+  <dt>Author</dt>
+  <dd>Jorge Olivero</dd>
+
+  <dt>Implements</dt>
+  <dd><a href="../rfcs/0001-subgraph-composition.md">RFC-0001 Subgraph Composition</a></dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/5">Engineering Plan PR</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-09</dd>
+
+  <dt>Date of approval</dt>
+  <dd>TBD</dd>
+
+  <dt>Approved by</dt>
+  <dd>TBD</dd>
+</dl>
+
+## Summary
+
+Subgraph composition allows a subgraph to import types from a foreign subgraph. Imports are designed to be loosely coupled throughout the deployment and indexing phases, and tightly coupled during query execution for the subgraph.
+
+To generate an API schema for the subgraph, the local schema needs to be merged with each of the foreign schemas that it imports. The merging process needs to take the following into account:
+
+1. An imported schema not being available on the graph-node
+2. An imported schema without a definition for the type which the local subgraph imports
+3. A merged and cached api schema for a subgraph imported by name is updated
+4. Ability to tell which subgraph/schema each type belongs to
+
+## Implementation
+
+There are two important parts to schema merging:
+
+1. Merging the schema
+
+2. Invalidating a cache entry when a subgraph's dependency, referenced by name, is updated and remerging that schema
+
+### Schema Merging
+
+Add a `merged_schema(&Schema, HashMap<SchemaReference, Arc<Schema>) -> Schema` function to the `graphql::schema` crate which will add each of the imported 
+types to the proviced document with a @subgraphId diretive denoting which subgraph the type came from. The `api_schema` function will add all the necessary types and fields for the imported types without any changes.
```

Comment:
As we established in the call about importing/implementing interfaces, just `@subgraphId` will not be enough. We also need to store the original type names, because imports may be aliased. So I suggest a new `@importInfo`, `@subgraphInfo` or `@mergeInfo` directive.